pkg/git/github.go

```diff
@@ -0,0 +1,307 @@
+			err = g.MergePR(ctx, int(pr.GetNumber()), *pr.GetHead().SHA)
+		}
+	}
+	return err
+}
+
+func (g *GitHubGITProvider) GetStatus(ctx context.Context, sha string, group string, env string) (Status, error) {
+	opts := &github.ListOptions{PerPage: 50}
+	statuses, _, err := g.client.Repositories.ListStatuses(ctx, g.owner, g.repo, sha, opts)
+	if err != nil {
+		return Status{}, err
+	}
+
+	name := fmt.Sprintf("%s-%s", group, env)
+	for _, s := range statuses {
+		comp := strings.Split(*s.Context, "/")
+		if len(comp) != 2 {
+			return Status{}, fmt.Errorf("status context in wrong format: %q", *s.Context)
+		}
+		if comp[1] == name {
+			return Status{
+				Succeeded: *s.State == "success",
+			}, nil
+		}
+	}
+
+	return Status{}, fmt.Errorf("no status found for sha %q", sha)
+}
+
+func (g *GitHubGITProvider) SetStatus(ctx context.Context, sha string, group string, env string, succeeded bool) error {
+	description := fmt.Sprintf("%s-%s-%s", group, env, sha)
+	name := fmt.Sprintf("kind/%s-%s", group, env)
+
+	state := "success"
+	if !succeeded {
+		state = "failure"
+	}
+
+	status := &github.RepoStatus{
+		State:       &state,
+		Context:     &name,
+		Description: &description,
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	_, _, err := g.client.Repositories.CreateStatus(ctx, g.owner, g.repo, sha, status)
+	return err
+}
+
+func (g *GitHubGITProvider) MergePR(ctx context.Context, id int, sha string) error {
+	opts := &github.PullRequestOptions{
+		SHA: sha,
+	}
+
+	var result *github.PullRequestMergeResult
+	var res *github.Response
+	err := retry.Do(
+		func() error {
+			var err error
+			result, res, err = g.client.PullRequests.Merge(ctx, g.owner, g.repo, id, "", opts)
+			if err != nil && res.StatusCode == 405 {
+				updateOpts := &github.PullRequestBranchUpdateOptions{}
+				_, _, err = g.client.PullRequests.UpdateBranch(ctx, g.owner, g.repo, id, updateOpts)
+			}
+			return err
+		},
+		retry.Attempts(5),
+		retry.LastErrorOnly(true),
+	)
+	if err != nil {
+		return err
+	}
+
+	mergeSucceeded := *result.Merged
+
+	if !mergeSucceeded {
+		body, err := ioutil.ReadAll(res.Response.Body)
+		if err != nil {
+			return err
+		}
+
+		defer func() {
+			err := res.Response.Body.Close()
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "MergePR - unable to close body: %v", err)
+			}
+		}()
+
+		return fmt.Errorf("PR with ID %d was not merged: %s", id, body)
+	}
+
+	return nil
+}
+
+func (g *GitHubGITProvider) GetPRWithBranch(ctx context.Context, source, target string) (PullRequest, error) {
+	listOpts := &github.PullRequestListOptions{
+		State: "open",
+		Base:  target,
+	}
+
+	var prs []*github.PullRequest
+	err := retry.Do(
+		func() error {
+			openPrs, _, err := g.client.PullRequests.List(ctx, g.owner, g.repo, listOpts)
+			if err != nil {
+				return err
+			}
+			for _, pr := range openPrs {
+				if source == *pr.Head.Ref {
+					prs = append(prs, pr)
+				}
+			}
+			if len(prs) == 0 {
```

Comment:
Should we do `len(prs) != 1` ?

Reply:
It should not be possible to have more than 1 PR, so that's fair.